loader/src/reference_binder.py

```diff
@@ -140,68 +93,60 @@ def bind(ref, isArray=False):
-                source_ref = (fhir_object["resourceType"], reference_path, isArray)
-                self.cache.sadd(
-                    json.dumps(target_ref),
-                    json.dumps((source_ref, fhir_object["id"]))
-                )
+                target_ref = self.identifier_to_key(reference_type, identifier)
+                source_ref = (fhir_object["resourceType"], reference_path, is_array)
+                self.cache.sadd(target_ref, json.dumps((source_ref, fhir_object["id"])))
             return ref
 
         # If we have a list of references, we want to bind all of them.
         # Thus, we loop on all the items in reference_attribute.
         if isinstance(reference_attribute, list):
-            return [bind(ref, isArray=True) for ref in reference_attribute]
+            return [bind(ref, is_array=True) for ref in reference_attribute]
         else:
             return bind(reference_attribute)
 
     @Timer("time_resolve_pending_references", "time spent resolving pending references")
     def resolve_pending_references(self, fhir_object):
         for identifier in fhir_object["identifier"]:
-            try:
-                identifier_tuple = self.extract_key_tuple(identifier)
-            except Exception as e:
-                logger.error(e)
-                continue
-            target_ref = json.dumps((fhir_object["resourceType"], identifier_tuple))
+            target_ref = self.identifier_to_key(fhir_object["resourceType"], identifier)
             pending_refs = self.load_cached_references(target_ref)
             for (source_type, reference_path, is_array), refs in pending_refs.items():
-                find_predicate = build_find_predicate(refs, reference_path, identifier, is_array)
-                update_predicate = build_update_predicate(reference_path, fhir_object, is_array)
+                find_predicate = self.build_find_predicate(
+                    refs,
+                    reference_path,
+                    identifier,
+                    is_array
+                )
+                update_predicate = self.build_update_predicate(
+                    reference_path,
+                    fhir_object,
+                    is_array
+                )
                 logger.debug(
                     f"Updating resource {source_type}: {find_predicate} {update_predicate}",
                     extra={"resource_id": get_resource_id(fhir_object)},
                 )
-                self.fhirstore.db[source_type].update_many(find_predicate, update_predicate)
+                if is_array:
+                    self.fhirstore.db[source_type].update_many(
+                        find_predicate,
+                        update_predicate,
+                        array_filters=[
+                            {
+                                "ref.identifier.value": identifier.get("value"),
+                                "ref.identifier.system": identifier.get("system")
```

Comment:
Oh, is this where the `$[ref]` is used?
Does this fix the update of several elements in an array? If so, could you add a test for that?

Reply:
It does. But I don't get how we could have several references with the same identifier.